[decisions] errors (#4716)
ElektraInitiative/libelektra: doc/decisions/0_drafts/errors_plugins.md

# Error Semantics

## Problem

While we have a classification of errors and warnings, it remains unclear when plugins actually should emit errors and warnings.

## Constraints

- Should not be contradicting to specified behavior in [storage plugin tutorial](/doc/tutorials/storage-plugins.md).

## Assumptions

- Users want a uniform behavior within Elektra, so plugins must behave uniformly.

## Considered Alternatives

### Free-for-all

Every plugin author decides for themselves.

Leads to inconsistent behaviour and hard to diagnose interaction issues.

### Detailed ruleset with conformance tests

Write a detailed formal ruleset and generic conformance tests for all plugins.

Defining the ruleset, let alone writing the tests, for all possible errors in plugins is borderline impossible.

> ### Simple guidelines

Reviewer: I fully agree with this decision. You can directly place the guidelines to `doc/contrib` in the next decision round.

Author: With "next round" do you mean the next PR, once this one is merged, or still within this PR?

Reviewer: Hard to say. Not so ideal that these two decisions are in one PR.

Author: > Not so ideal that these two decisions are in one PR.

I didn't expect this much discussion... I thought the decisions are both very straightforward.

Create simple guidelines for plugin authors and clearly document them.
Rely on plugin authors adhering to the guidelines and testing their plugins themselves.
Plugins that do not adhere to the guidelines can be marked down via `infos/status`.

The basic guidelines will be (but may be extended):

1. During `kdbGet` only emit warnings, unless the error is entirely unrecoverable:

>    Whatever other guidelines say, during `kdbGet` we plugins shall only emit warnings to allow the caller to fix issues via `kdbSet`.

Reviewer: ```suggestion
   Whatever other guidelines say, during `kdbGet` plugins shall only emit warnings to allow the caller to fix issues via `kdbSet`.
   If possible, the emitted warnings should be identical to the errors set in `kdbSet()`.
```

>    Only if errors are entirely unrecoverable, should an error be emitted.

Reviewer: ```suggestion
   Only if errors are entirely unrecoverable, an error should be emitted.
```

   Some examples for either case:

   1. File-based backend cannot parse local file:
      Emit a warning.
      We may still be able to overwrite the file in `kdbSet`.
   2. File-based backend cannot access file, because path is invalid (e.g., on Windows `C:\foo\CON`)
      Emit an error.
      A `kdbSet` could recover, but the mountpoint definition needs to change.
      To do this, a new `kdbOpen` is needed.

2. Prefer errors to warnings in `kdbSet`:
   Most validation detect either valid or invalid values.

>    Only in the rare case, where something is valid, but might not be correct should a warning be used.

Reviewer: Can you please give an example?

```suggestion
   Only in the rare case, where something is valid, but might not be correct a warning should be used.
```

   For very common mistakes which are still technically valid configuration, both emitting an error and emitting a warning are valid options.

>    But in both cases, there should be a way to suppress or avoid the error/warning.

Reviewer: I wouldn't recommend that as it would require config options to most plugins. We should be careful with such warnings, if we even need them (looking forward to examples).

Author: Maybe not phrased well, but it doesn't have to be a config option. And IMO a config option would be a bad choice because it is too global. It could simply be an additional metakey, or maybe a different value, to indicate you know what you're doing.

One example (which I will mention in the file) I know for a case like this is in the new `backend` plugin:

https://github.com/ElektraInitiative/libelektra/blob/8fe1340ea7f7bc37aa2f14bc0d331d954cf0e216/src/plugins/backend/backend.c#L235-L243

Technically a correct configuration, but in most cases probably not what the user intended. If they did want to do it this way, there is a clear way to suppress the warning.

Other cases could maybe be in the `spec` plugin. For example, if we do add a check to protect against mixed array and non-array children. There could be an exotic use case where this is still needed. It would make sense to allow adding something like `meta:/array/mixed = allow` to the array parent `spec:/foo/#` to avoid the warnings on every `kdbGet` and to allow a `kdbSet`.

3. Don't stop after the first error:
   Additional errors, are automatically converted into warnings.
   Report as many errors as possible in one go, so the user can fix as much as possible in one go.

4. Specifications are exact:
   Both the checked config must conform **exactly** to the specification.
   Plugins should not ignore unknown values or values they don't understand.

5. Reject unknowns:

>    To be on the safe side, always reject any unknowns that are direct at the plugin.

Reviewer: "direct at"?

Author: Should be "directed at"

```suggestion
   To be on the safe side, always reject any unknowns that are directed at the plugin.
```

I needed a general term for "things the plugin should handle". Not all plugins are triggered by presence of metadata. Some (like storage plugins) try to handle everything, others get a parent from somewhere, etc.

   Unknown specification should always result in errors.
   Unknown config data may result in warnings instead, unless it is known to be invalid.
   This mostly applies to storage plugins.
   For example, if a plugin handles `meta:/foo`, it should emit an error, if encounters an unknown value for `meta:/foo`.

   Of course, plugins can decide that they are not handling certain keys or metakeys.
   But that should only be based on the keyname, never the value.

6. The `storage` phase should handle everything during `kdbSet`:
   A plugin executed in the `storage` phase of `kdbSet` should store the entire `KeySet` they receive, such that in a future `kdbGet` they will load an equal `KeySet`.
   If such a plugin cannot handle the entire `KeySet` semantics (e.g., no metadata), the unsupported data should be removed in an earlier phase.

>    Either by the plugin itself, or by another plugin.

Reviewer: ```suggestion
   Either by the plugin itself, or by another plugin that is required by said plugin.
```
